projection/ibc_channel/ibc_channel_test.go

```diff
@@ -0,0 +1,828 @@
+				mockTx := NewMockRDbTx()
+				mocks = append(mocks, &mockTx.Mock)
+				mockCoon.On("Begin").Return(mockTx, nil)
+
+				mockExecResult := &test.MockRDbExecResult{}
+				mocks = append(mocks, &mockExecResult.Mock)
+				mockExecResult.On("RowsAffected").Return(int64(1))
+				mockTx.On(
+					"Exec",
+					"UPDATE view_ibc_channels SET total_transfer_out_success_count = total_transfer_out_success_count+1 WHERE channel_id = $1",
+					"SourceChannel",
+				).Return(mockExecResult, nil)
+
+				mockTx.On(
+					"Exec",
+					"UPDATE view_ibc_channels SET last_out_packet_sequence = $1 WHERE channel_id = $2",
+					uint64(1),
+					"SourceChannel",
+				).Return(mockExecResult, nil)
+
+				mockTx.On(
+					"Exec",
+					"UPDATE view_ibc_channels SET last_activity_block_height = $1, last_activity_block_time = $2 WHERE channel_id = $3",
+					int64(1),
+					int64(0),
+					"SourceChannel",
+				).Return(mockExecResult, nil)
+
+				mockRowResult := &test.MockRDbRowResult{}
+				mocks = append(mocks, &mockRowResult.Mock)
+				var totalTransferOutCount int64
+				var totalTransferOutSuccessCount int64
+				mockRowResult.On("Scan", &totalTransferOutCount, &totalTransferOutSuccessCount).Return(nil).Run(func(args testify_mock.Arguments) {
+					totalTransferOutCount := args.Get(0).(*int64)
+					totalTransferOutSuccessCount := args.Get(1).(*int64)
+					*totalTransferOutCount = 1
+					*totalTransferOutSuccessCount = 1
+				})
+				mockTx.On(
+					"QueryRow",
+					"SELECT total_transfer_out_count, total_transfer_out_success_count FROM view_ibc_channels WHERE channel_id = $1",
+					"SourceChannel",
+				).Return(mockRowResult, nil)
+
+				var bondedTokensJSON string
+				mockRowResult.On("Scan", &bondedTokensJSON).Return(nil).Run(func(args testify_mock.Arguments) {
+					bondedTokensJSON := args.Get(0).(*string)
+					*bondedTokensJSON = "{\"test\":\"\"}"
+				})
+				mockTx.On(
+					"QueryRow",
+					"SELECT bonded_tokens FROM view_ibc_channels WHERE channel_id = $1",
+					"SourceChannel",
+				).Return(mockRowResult, nil)
+
+				mockTx.On(
+					"Exec",
+					"UPDATE view_ibc_channels SET total_transfer_out_success_rate = $1 WHERE channel_id = $2",
+					float64(1),
+					"SourceChannel",
+				).Return(mockExecResult, nil)
+
+				mockTx.On(
+					"Exec",
+					"UPDATE view_ibc_channels SET bonded_tokens = $1 WHERE channel_id = $2",
+					"{\"onThisChain\":null,\"onCounterpartyChain\":[{\"denom\":\"//DENOM\",\"amount\":\"100\"}]}",
+					"SourceChannel",
+				).Return(mockExecResult, nil)
+
+				mockUpdateLastHandledEventHeight(mockTx)
+
+				return mocks
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		mockRDbConn := NewMockRDbConn()
+		mocks := tc.MockFunc(mockRDbConn)
+		mocks = append(mocks, &mockRDbConn.Mock)
+
+		projection := NewIBCChannelProjection(mockRDbConn)
+		err := projection.HandleEvents(1, tc.Events)
+		assert.NoError(t, err)
+
+		for _, m := range mocks {
+			m.AssertExpectations(t)
```

Comment:
May I know what will be assert upon running this line? Is that all `mockTx.On(...)` is expected to be run at least once?

Reply:
Got it, one critics of mocking the RDb would be that the test will be very dependant on the implementation but what we concern about is more on the outcome.

Let's say one day we have optimized the handling logic, or do single insert all instead of multiple insertions, this will break the test and is hard to update.

Reply:
Overall I like the framework and I also saw it in many other Golang projects which I think is clear. But what do assert I think we can think about it.

Reply:
I see your concern. There is a `Repeatability` for each call (if same method name and same arguments). Test cases can use `.Once()`, `.Tiwce()` or` Times(int)` to indicated it. By default, the assertion would already pass if no `Repeatability` was specified.